src/ws/mask.rs

```diff
@@ -1,89 +1,107 @@
-        let head = min(len, (8 - (ptr as usize & 0x7)) & 0x3);
-        let mask_u32 = if head > 0 {
-            let n = if head > 4 { head - 4 } else { head };
-
-            let mask_u32 = if n > 0 {
-                xor_mem(ptr, mask_u32, n);
-                ptr = ptr.offset(head as isize);
-                len -= n;
-                if cfg!(target_endian = "big") {
-                    mask_u32.rotate_left(8 * n as u32)
-                } else {
-                    mask_u32.rotate_right(8 * n as u32)
-                }
-            } else {
-                mask_u32
-            };
-
-            if head > 4 {
-                *(ptr as *mut u32) ^= mask_u32;
-                ptr = ptr.offset(4);
-                len -= 4;
-            }
-            mask_u32
+    // Extend the mask to 64 bits
+    let mut mask_u64 = ((mask_u32 as u64) << 32) | (mask_u32 as u64);
+    // Split the buffer into three segments
+    let (head, mid, tail) = align_buf(buf);
+
+    // Initial unaligned segment
+    let head_len = head.len();
+    if head_len > 0 {
+        xor_short(head, mask_u64);
+        if cfg!(target_endian = "big") {
+            mask_u64 = mask_u64.rotate_left(8 * head_len as u32);
         } else {
-            mask_u32
-        };
-
-        if len > 0 {
-            debug_assert_eq!(ptr as usize % 4, 0);
+            mask_u64 = mask_u64.rotate_right(8 * head_len as u32);
         }
+    }
+    // Aligned segment
+    for v in mid {
+        *v ^= mask_u64;
+    }
+    // Final unaligned segment
+    if tail.len() > 0 {
+        xor_short(tail, mask_u64);
+    }
+}
 
-        // Properly aligned middle of the data.
-        if len >= 8 {
-            let mut mask_u64 = mask_u32 as u64;
-            mask_u64 = mask_u64 << 32 | mask_u32 as u64;
-
-            while len >= 8 {
-                *(ptr as *mut u64) ^= mask_u64;
-                ptr = ptr.offset(8);
-                len -= 8;
-            }
-        }
+#[inline]
+// TODO: copy_nonoverlapping here compiles to call memcpy. While it is not so
+// inefficient, it could be done better. The compiler does not understand that
+// a `ShortSlice` must be smaller than a u64.
+fn xor_short(buf: ShortSlice, mask: u64) {
+    // Unsafe: we know that a `ShortSlice` fits in a u64
+    unsafe {
+        let (ptr, len) = (buf.0.as_mut_ptr(), buf.0.len());
+        let mut b: u64 = 0;
+        #[allow(trivial_casts)]
+        copy_nonoverlapping(ptr, &mut b as *mut _ as *mut u8, len);
+        b ^= mask;
+        #[allow(trivial_casts)]
+        copy_nonoverlapping(&b as *const _ as *const u8, ptr, len);
+    }
+}
 
-        while len >= 4 {
-            *(ptr as *mut u32) ^= mask_u32;
-            ptr = ptr.offset(4);
-            len -= 4;
-        }
+#[inline]
+// Unsafe: caller must ensure the buffer has the correct size and alignment
+unsafe fn cast_slice(buf: &mut [u8]) -> &mut [u64] {
+    // Assert correct size and alignment in debug builds
+    debug_assert!(buf.len() & 0x7 == 0);
```

Comment:
Emm just to clarify does this mean that our buffer here must be at least aligned and with size `8` or any other that `len % 8 == 0`?

I feel like it would be better to have comment here for clarification